libyul/optimiser/DeadCodeEliminator.h

```diff
@@ -0,0 +1,52 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that removes unused variables and functions.
+ */
+
+#pragma once
+
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/YulString.h>
+
+#include <map>
+#include <set>
+
+namespace yul
+{
+struct Dialect;
+
+/**
+ * Optimisation stage that removes unreachable code
+ *
+ * Unreachable code is any code within a block which is preceded by an
+ * unconditional return, break, continue, selfdestruct or revert.
```

Comment:
I wanted to differentiate between a `return` that is in the body-block of an `if`

Reply:
I think "unconditional" is not really the right word here and also sounds like it would be a semantic property instead of just a purely syntactic. The component should not modify the following code (please add a test for that), even though the `mstore` follows an unconditional revert.
```
{
  revert(0, 0)
}
mstore(0, 0)
```

Please also mention that function definitions are not removed.